distributed/platform.py

```diff
@@ -0,0 +1,70 @@
+import os
+import sys
+
+import psutil
+
+__all__ = ("memory_limit", "cpu_count", "PLATFORM_MEMORY_LIMIT", "PLATFORM_CPU_COUNT")
+
+
+def memory_limit():
+    """Get the memory limit (in bytes) for this system.
+
+    Takes the minimum value from the following locations:
+
+    - Total system host memory
+    - Cgroups limit (if set)
+    - RSS rlimit (if set)
+    """
+    limit = psutil.virtual_memory().total
+
+    # Check cgroups if available
+    if sys.platform == "linux":
+        try:
+            with open("/sys/fs/cgroup/memory/memory.limit_in_bytes") as f:
+                cgroups_limit = int(f.read())
+            if cgroups_limit > 0:
+                limit = min(limit, cgroups_limit)
+        except Exception:
+            pass
+
+    # Check rlimit if available
+    try:
+        import resource
+
+        hard_limit = resource.getrlimit(resource.RLIMIT_RSS)[1]
+        if hard_limit > 0:
+            limit = min(limit, hard_limit)
+    except (ImportError, OSError):
+        pass
+
+    return limit
+
+
+def cpu_count():
+    """Get the available CPU count for this system.
+
+    Takes the minimum value from the following locations:
+
+    - Total system cpus available on the host.
+    - Cgroups limit (if set)
+    """
+    count = os.cpu_count()
```

Comment:
Since `psutil` is a dep could we use it in preference here as it's more robust that the Python implementation.

In particular I'm running into a Python bug (win64/py37) whereby it thinks I have 64 cores when my VMWare VM only has 4 VCPUs allocated:
```python
In [3]: os.cpu_count()
Out[3]: 64

In [4]: psutil.cpu_count()
Out[4]: 4
```

Furthermore, the number of cores on the system isn't representative of the number of cores available to the process.  Docker containers are only one example of this. On some HPC systems it's usual for processes to have their affinity set to lock them to particular cores. Since we're making improvements here to consider the number of cores *available* I think it would be a useful improvement to use the cpu affinity rather than the cpu count:
```python
In [5]: psutil.Process().cpu_affinity()
Out[5]: [0, 1, 2, 3]

In [6]: len(psutil.Process().cpu_affinity())
Out[6]: 4
```
```suggestion
    count = len(psutil.Process().cpu_affinity())
```

***xref:*** https://bugs.python.org/issue26692#msg310113

Reply:
Cool, good to know. I've added this to the set of things we check, but we can't rely on it 100% since it's not available on all platforms.

Out of curiousity, does `os.sched_getaffinity(0)` work for your use case?